### ISSUE TITLE
Mouse Capturing Problems

### ISSUE THREAD

**Issue by Flyxi121** [gnif/LookingGlass#501]
When I move the mouse to the left side of the screen and click (flicking in a shooter game) in capture mode, looking glass does not let me click anymore. After that happens, I don't see a mouse pointer in windows, only a small dot, which is slightly to the right of the real mouse pointer in windows.
I don't know how I would show you any debug information about the mouse clicking and mouse tracking.
I'm using sway (wayland).

**Comment by Xyene** [comment]
What version of Looking Glass are you using, and what version of sway/wlroots?

Could start Looking Glass with `WAYLAND_DEBUG=1 looking-glass-client [your params] |& tee /tmp/log` and share the resulting log (after reproducing the issue in as short a time as possible, to keep the log small)?

**Comment by Flyxi121** [comment]
I'm using the B3 release of looking glass and sway 1.5.1 and wlroots 0.12.0.

The log after tabbing into the window, reproducing the issue (about 5 seconds) and closing it again is about 12000 lines:
[log.txt](https://github.com/gnif/LookingGlass/files/6058283/log.txt)

**Comment by Xyene** [comment]
Thanks for the logs. The problematic bit is:

```
[3749433.879] zwp_relative_pointer_v1@45.relative_motion(8, 2875293632, -24.000000, -2.000000, -12.000000, -1.000000)
[3749433.898] wl_pointer@36.motion(37235032, 0.000000, 916.070312)
[3749433.903] wl_pointer@18.motion(37235032, 0.000000, 916.070312)
[3749433.908] wl_pointer@18.frame()

--- I think your click happened here ---

[3749433.909] wl_pointer@36.leave(35147, wl_surface@22)
[3749433.912] wl_pointer@18.leave(35147, wl_surface@22)
[3749433.917]  -> wl_pointer@18.set_cursor(0, wl_surface@40, 4, 0)
[3749433.922]  -> wl_surface@40.attach(wl_buffer@39, 0, 0)
[3749433.926]  -> wl_surface@40.damage(0, 0, 8, 4)
[3749433.930]  -> wl_surface@40.commit()
[3749433.932]  -> wl_pointer@18.set_cursor(0, wl_surface@20, 5, 5)
[3749433.936]  -> wl_surface@20.attach(wl_buffer@3, 0, 0)
[3749433.940]  -> wl_surface@20.damage(0, 0, 32, 32)
```

The pointer confine isn't destroyed before that so that leave should not have happened. Perhaps sway is getting being confused by the fact that we have two pointer instances in B3 (one for SDL and one for Looking Glass itself)... at any rate, this is technically legal behavior according to the constraint spec.

Could you try Looking Glass master and see if it reproduces there? Wayland is substantially better there, as we've gotten rid of SDL entirely. It's possible that the bug won't occur there, or behave differently if it does.

**Comment by Flyxi121** [comment]
Okay now with the bleeding edge version the problem is a bit different. Sometimes it does the same as before but most of the time now the mouse just leaves the window entirely when I move the mouse to the left and click (with capturing of course).

This is the new log:
[log_bleeding-edge.txt](https://github.com/gnif/LookingGlass/files/6058430/log.txt)

**Comment by Xyene** [comment]
Thanks for trying. The log is a bit more clear now:

```
[1411199.212] wl_pointer@17.motion(39191766, 0.000000, 586.562500)
[1411200.187] zwp_relative_pointer_v1@3.relative_motion(9, 537061336, -42.000000, 0.000000, -21.000000, 0.000000)
[1411201.182] zwp_relative_pointer_v1@3.relative_motion(9, 537062336, -46.000000, 0.000000, -23.000000, 0.000000)
[1411202.181] zwp_relative_pointer_v1@3.relative_motion(9, 537063336, -38.000000, 0.000000, -19.000000, 0.000000)
[1411203.190] zwp_relative_pointer_v1@3.relative_motion(9, 537064336, -48.000000, 0.000000, -24.000000, 0.000000)

--- click happens here? ---

[1411203.202] wl_pointer@17.leave(40024, wl_surface@19)
[1411203.206]  -> wl_pointer@17.set_cursor(40019, wl_surface@25, 0, 0)
[1411203.211]  -> zwp_confined_pointer_v1@35.destroy()
```

Some more questions:

* I assume you don't have something like `seat ... pointer_constraint disable` in your sway config?
* Does this only happen on the left edge of the window?
* What's your monitor layout? Are any of your monitors in negative coordinate space?
* If you compile https://github.com/swaywm/wlroots/blob/master/examples/pointer-constraints.c, run `./pointer-constraints confine persistent no-region` and fullscreen it, can you reproduce with that? Note that `pointer-constraints` toggles the constraint area on click, so testing with the top left corner of the window is probably easiest.

**Comment by Flyxi121** [comment]
To answer your questions:
- No I do not have `seat ... pointer_constraint disable` in my config
- Yes, it only happens on the left side
-  Looking glass on my main monitor and another one to left, no coordinates are negatives
- I can not reproduce the same issue with `./pointer-constraints confine persistent no-region` (I can not escape the fullscreen window)

**Comment by Xyene** [comment]
Thanks for the quick replies... this is pretty  strange.

Another question: does the issue reproduce if you fullscreen Looking Glass, then toggle the capture off / on again (as opposed to capturing and then fullscreening)? There's at least one bug (https://github.com/swaywm/sway/issues/6070) in sway that will cause misbehaviors in the first case.

Would you be willing to compile sway with a patch applied?

```diff
diff --git a/sway/input/cursor.c b/sway/input/cursor.c
index cbb5c6e9..2552bcbf 100644
--- a/sway/input/cursor.c
+++ b/sway/input/cursor.c
@@ -348,6 +348,9 @@ static void pointer_motion(struct sway_cursor *cursor, uint32_t time_msec,
 		cursor->seat->wlr_seat, (uint64_t)time_msec * 1000,
 		dx, dy, dx_unaccel, dy_unaccel);
 
+	sway_log(SWAY_ERROR, "CURSOR: entry at %f, %f",
+			cursor->cursor->x, cursor->cursor->y);
+
 	// Only apply pointer constraints to real pointer input.
 	if (cursor->active_constraint && device->type == WLR_INPUT_DEVICE_POINTER) {
 		struct wlr_surface *surface = NULL;
@@ -356,20 +359,39 @@ static void pointer_motion(struct sway_cursor *cursor, uint32_t time_msec,
 			cursor->cursor->x, cursor->cursor->y, &surface, &sx, &sy);
 
 		if (cursor->active_constraint->surface != surface) {
+			sway_log(SWAY_ERROR, "CONSTRAINT: surface mismatch");
 			return;
 		}
 
 		double sx_confined, sy_confined;
 		if (!wlr_region_confine(&cursor->confine, sx, sy, sx + dx, sy + dy,
 				&sx_confined, &sy_confined)) {
+			sway_log(SWAY_ERROR, "CONSTRAINT: bail");
 			return;
 		}
 
 		dx = sx_confined - sx;
 		dy = sy_confined - sy;
+
+		wlr_cursor_move(cursor->cursor, device, dx, dy);
+
+		struct wlr_surface *psurface = NULL;
+		double psx, psy;
+		node_at_coords(cursor->seat,
+			cursor->cursor->x, cursor->cursor->y, &psurface, &psx, &psy);
+		sway_log(SWAY_ERROR,
+				"CONSTRAINT: dx=%f, dy=%f, sx=%f, sy=%f, surface=%p, psx=%f, psy=%f, psurface=%p",
+				dx, dy, sx, sy, surface, psx, psy, psurface);
+
+		if (surface != psurface) {
+			abort();
+		}
+	} else {
+		wlr_cursor_move(cursor->cursor, device, dx, dy);
 	}
 
-	wlr_cursor_move(cursor->cursor, device, dx, dy);
+	sway_log(SWAY_ERROR, "CURSOR: post-warp at %f, %f",
+			cursor->cursor->x, cursor->cursor->y);
 
 	seatop_pointer_motion(cursor->seat, time_msec);
 }
@@ -408,6 +430,14 @@ void dispatch_cursor_button(struct sway_cursor *cursor,
 		time_msec = get_current_time_msec();
 	}
 
+	struct wlr_surface *surface = NULL;
+	double sx, sy;
+	node_at_coords(cursor->seat,
+		cursor->cursor->x, cursor->cursor->y, &surface, &sx, &sy);
+	sway_log(SWAY_ERROR,
+			"CLICK: cx=%f, cy=%f, sx=%f, sy=%f, surface=%p",
+			cursor->cursor->x, cursor->cursor->y, sx, sy, surface);
+
 	seatop_button(cursor->seat, time_msec, device, button, state);
 }
```

You can find instructions for compiling sway from source here: https://github.com/swaywm/sway/wiki/Development-Setup

I'd be interested to see sway's log while you reproduce the issue with the patch applied. I think this may be a sway-side bug.

**Comment by Flyxi121** [comment]
Okay so updating to bleeding edge definetly stopped the issue from happening 99% of the time. After shutting down everything yesterday and booting my system up this morning, my mouse only rarely escapes the window (I managed to do it once in about an hour of gaming). Then I wondered if the issue would also be resolved with release B3 and downgraded, but it was not. 
As the issue did not happen anymore with the newest version I tried the "fullscreen and then capture on and off" with B3 but it did not change anything.
I will try the sway patch with looking-glass B3 later that day, reproduce the issue and show you sway's log.
Thanks for your great help!

**Comment by Xyene** [comment]
>  The mouse also escapes the window on the bleeding-edge version when there are other windows behind looking-glass.

This makes it sound more and more like a compositor bug. Can you take some screenshots of the kind of layout you're experiencing this with? Your exact `output` definitions in your config might help me reproduce it on my end.

Also, could you give https://github.com/gnif/LookingGlass/pull/504 a try? It won't fix sway, but it ought to fix the symptoms in Looking Glass. In the case it does, I'd still like to diagnose the bug in sway -- I just don't think I'd have time to get it in before the 1.6 release there, so the "workaround" in Looking Glass makes sense.

**Comment by Flyxi121** [comment]
As far as I can say at the moment, the issue is not happening anymore with the `cursor.c` patch you asked me to test.

My layout is a bit of a weird one, as I have 2 ultra-wide screens and one of them is vertical. This is the output definition in the sway config:
```
output $monitor_right pos 1440 1050 res 3440x1440@100Hz
output $monitor_left pos 0 0 res 3440x1440@60Hz transform 270
```
And this is a screenshot of both screens:
![image](https://user-images.githubusercontent.com/53060619/109646113-0cfa4980-7b58-11eb-8f58-aee5b6d68350.png)

I will update you shortly about #504.

**Comment by Xyene** [comment]
> Looking-glass bleeding edge still lets the mouse escape the window but only when something it behind the fullscreen client

What exactly does this mean? Are the windows "behind" the fullscreen window in tiling mode, or floating mode? Does it matter? Does it matter where the windows are?

> The sway `input/cursor.c` patch prevents the issue from happening at all

For the avoidance of doubt: I imagine you applied the patch on sway master. Did you try reproducing on sway master, without the patch applied? I'm asking because that patch should have been purely logging, and not contain any functional difference. It's possible the bug could have been resolved since 1.5.1 -- in particular, https://github.com/swaywm/sway/commit/0df5753aed57c6b5e3a5f64ff410cadf8d96c35c comes to mind.

> Patch #504 also prevents the issue from happening at all

Thanks for checking; that patch makes the cursor avoid edges in the first place :)

**Comment by Flyxi121** [comment]
> What exactly does this mean? Are the windows "behind" the fullscreen window in tiling mode, or floating mode? Does it matter? Does it matter where the windows are?
- There has to be at least one window in tiling mode behind in tiling mode and I could not reproduce the issue with the window was in floating or fullscreen mode.
- The location did not matter and there does not have to be anything on the other monitor.

> For the avoidance of doubt: I imagine you applied the patch on sway master. Did you try reproducing on sway master, without the patch applied? I'm asking because that patch should have been purely logging, and not contain any functional difference. It's possible the bug could have been resolved since 1.5.1 -- in particular, swaywm/sway@0df5753 comes to mind.

Yeah I was also wondering that the "debug-patch"  changed anything but I did not realize that I used sway master.
I have tried sway master without the patch and obviously the patch was not the changing factor but the master version was. I have also tried to reproduce the issue with sway-master and looking-glass B3 and I also could not reproduce the issue.

So in the end everything is fixed by just using sway-master as it seems to be a compositor related issue.
